### ISSUE TITLE
Feedback on Difficulty factor

### ISSUE THREAD

**Issue by royragsdale** [robinsloan/spring-83#8]
# Issue

The [difficulty factor](https://github.com/robinsloan/spring-83-spec/blob/main/draft-20220616.md#difficulty-factor) seems ineffective in making keys harder to find as in practice only the most significant bits (e.g. the first "hex character") dominate.

## Additional spec wording feedback

"The key's first 16 hex characters, interpreted as a 64-bit number". I assumed interpreted as a big endian, unsigned, integer. If retained, likely worth making this explicit.

The expected precision and rounding behavior of the float math is not defined. In general the use of float/fractions introduces a lot of likely unnecessary complexity.

## Recommendation

I unfortunately don't have a replacement difficulty factor to recommend but wanted to point out the behavior I was observing (e.g. it's easy to generate "strong" keys). For future iterations recommend more precise working on numeric interpretations, the removal of floats, and a table of fully calculated tests/examples (e.g. difficulty factor of X = key threshold Y, the test key has a strength of Z).

## Supporting Data

Here are three example that demonstrate the issue, generated with my work in progress [client](https://github.com/royragsdale/s83/), which I believe is correctly generating thresholds. All keys were essentially equivalently difficult to generate, computed in seconds/minutes.  Matching the key format `83eMMYY$` is the more restrictive/hard part.  The specific times to generate are not overly relevant, more just to demonstrate the dominance of the most significant bits (e.g. 1/16 of keys will randomly lead with a "0" and meet a difficulty factor >0.9).

A. 
```
$ ./s83 new -j 12
...
[info] Success! Found a valid key in 14654732 iterations over 91 seconds (159770 kps)
[info] This key passes a difficulty factor of ~0.362 (strength=11756531408393154395)
...
public: a3279706f3ff2f5bbf2882fd86d9c2b58e9061a5edae345e435d116fd83e0124
```

B. 
```
$ ./s83 new -j 12
...
[info] Success! Found a valid key in 2149599 iterations over 13 seconds (158698 kps)
[info] This key passes a difficulty factor of ~0.879 (strength=2221055228358208723)
...
public: 1ed2c5f668d0acd30d0e8d117763067ab6d73232e1f4f189b1569ad7983e0722
```

C.
```
$ ./s83 new -j 12
...
[info] Success! Found a valid key in 36940562 iterations over 234 seconds (157278 kps)
[info] This key passes a difficulty factor of ~0.939 (strength=1120100156655163821)
...
public: 0f8b65286e4a59ad35b074520808c8dc6340aeb180fe03120cd7f614583e0323
```

**Comment by robinsloan** [comment]
This feedback is very well-taken; thank you for the investigation. As I've nudged the protocol toward something more "server-oriented", relenting on the "everything everywhere" design I started with, the difficulty factor has seemed suddenly less important. A server under stress can, after all… just close its doors to new publishers! Or return error code 429, or require some kind of extra authorization, or any of a number of other things, none of them this baroque.

So, that's all to say, I will probably drop the difficulty factor from the next draft, not without regret -- this kind of thing has such a cyberpunk-y appeal -- but, a "feature" that's (a) poorly specified, (b) probably not necessary, and (c) difficult to implement is a pretty good candidate for removal.
